File: kotlin/src/main/kotlin/com/murex/tbw/finance/CurrencyOperations.kt
/*******************************************************************************
 *
 * Copyright (c) {2003-2021} Murex S.A.S. and its affiliates.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the MIT License
 * which accompanies this distribution, and is available at
 * https://opensource.org/licenses/MIT
 *
 *******************************************************************************/

package com.murex.tbw.finance

import com.murex.tbw.domain.country.Currency
import com.murex.tbw.domain.country.Currency.*
import kotlin.math.floor

class CurrencyOperations {

    companion object Converter {
        private val EXCHANGE_RATE = mapOf(
            US_DOLLAR to 1.0,
            EURO to 1.14,
            POUND_STERLING to 1.27,
            RENMINBI to 0.15,
            YEN to 0.0093,
            AUSTRALIAN_DOLLAR to 0.7
        )

        @JvmStatic
        fun fromUSD(input: Double, toCurrency: Currency): Double {
            val value = input / (EXCHANGE_RATE[toCurrency] ?: error(""))
            return floor(value * 100) / 100
        }

        @JvmStatic
        fun toUSD(input: Double, fromCurrency: Currency): Double {
            val value = input * (EXCHANGE_RATE[fromCurrency] ?: error(""))
            return floor(value * 100) / 100
        }
    }
}

File: scala/src/main/scala/com/murex/tbw/finance/CurrencyOperations.scala
/*
 * ******************************************************************************
 *  *
 *  * Copyright (c) {2003-2022} Murex S.A.S. and its affiliates.
 *  * All rights reserved. This program and the accompanying materials
 *  * are made available under the terms of the MIT License
 *  * which accompanies this distribution, and is available at
 *  * https://opensource.org/licenses/MIT
 *  *
 *  ******************************************************************************
 */

package com.murex.tbw.finance

import com.murex.tbw.domain.country.Currencies._

import scala.math.floor

object CurrencyOperations {
  private val EXCHANGE_RATE = Map(
    US_DOLLAR -> 1.0,
    EURO -> 1.14,
    POUND_STERLING -> 1.27,
    RENMINBI -> 0.15,
    YEN -> 0.0093,
    AUSTRALIAN_DOLLAR -> 0.7
  )

  def fromUSD(input: Double, toCurrency: Currency): Double = {
    val value: Double =
      input / EXCHANGE_RATE.getOrElse(toCurrency, throw new IllegalStateException(""))
    floor(value * 100) / 100
  }

  def toUSD(input: Double, toCurrency: Currency): Double = {
    val value: Double =
      input * EXCHANGE_RATE.getOrElse(toCurrency, throw new IllegalStateException(""))
    floor(value * 100) / 100
  }
}
